internal/message/amqp/publisher_test.go

```diff
@@ -0,0 +1,319 @@
+
+	fn(mockAdapter, mockConn, mockChan)
+
+	origAdapter := defaultDialerAdapter
+	origRetryDelay := connectionRetryDelay
+
+	defaultDialerAdapter = mockAdapter.Dial
+	connectionRetryDelay = 1 * time.Nanosecond
+
+	fixture := &connectFixture{
+		adapter:    mockAdapter,
+		connection: mockConn,
+		channel:    mockChan,
+	}
+	reset := func() {
+		defaultDialerAdapter = origAdapter
+		connectionRetryDelay = origRetryDelay
+	}
+	return fixture, reset
+}
+
+func TestNewPublisher(t *testing.T) {
+	t.Run("connect", func(t *testing.T) {
+		f, reset := setupConnect(func(adapter *mockDialerAdapter, conn *mockConnection, channel *mockChannel) {
+			conn.On("Channel").Return(channel, nil)
+			adapter.On("Dial", uri).Return(conn, nil)
+		})
+		defer reset()
+
+		actual, err := NewPublisher(uri, queueName, logger)
+		require.NoError(t, err)
+		assert.NotNil(t, actual.conn)
+		assert.NotNil(t, actual.channel)
+		assert.Equal(t, queueName, actual.queueName)
+		assert.Equal(t, uri, actual.uri)
+
+		f.adapter.AssertExpectations(t)
+		f.connection.AssertExpectations(t)
+	})
+
+	t.Run("reconnect", func(t *testing.T) {
+		f, reset := setupConnect(func(adapter *mockDialerAdapter, conn *mockConnection, channel *mockChannel) {
+			conn.On("Channel").Return(channel, nil)
+			adapter.On("Dial", uri).Return(nil, errors.New("test dial error")).Once()
+			adapter.On("Dial", uri).Return(conn, nil).Once()
+		})
+		defer reset()
+
+		actual, err := NewPublisher(uri, queueName, logger)
+		require.NoError(t, err)
+		assert.NotNil(t, actual.conn)
+		assert.NotNil(t, actual.channel)
+		assert.Equal(t, queueName, actual.queueName)
+		assert.Equal(t, uri, actual.uri)
+
+		f.adapter.AssertExpectations(t)
+		f.adapter.AssertNumberOfCalls(t, "Dial", 2)
+		f.connection.AssertExpectations(t)
+	})
+
+	t.Run("channel_failure", func(t *testing.T) {
+		f, reset := setupConnect(func(adapter *mockDialerAdapter, conn *mockConnection, channel *mockChannel) {
+			conn.On("Channel").Return(nil, errors.New("test channel failure"))
+			adapter.On("Dial", uri).Return(conn, nil)
+		})
+		defer reset()
+
+		_, err := NewPublisher(uri, queueName, logger)
+		assert.Error(t, err)
+
+		f.adapter.AssertExpectations(t)
+		f.connection.AssertExpectations(t)
+	})
+
+	t.Run("retry_limit_failure", func(t *testing.T) {
+		f, reset := setupConnect(func(adapter *mockDialerAdapter, conn *mockConnection, channel *mockChannel) {
+			adapter.On("Dial", uri).Return(nil, errors.New("test dial error"))
+		})
+		defer reset()
+
+		_, err := NewPublisher(uri, queueName, logger)
+		assert.Error(t, err)
+
+		f.adapter.AssertExpectations(t)
+		f.adapter.AssertNumberOfCalls(t, "Dial", connectionRetryLimit)
+	})
+}
+
+func TestPublisher_Push(t *testing.T) {
+	testMessage := "hello"
+
+	t.Run("success", func(t *testing.T) {
+		mockChan := &mockChannel{}
+
+		mockChan.On("QueueDeclare", queueName, true, false, false, false, amqp.Table{
+			"x-single-active-consumer": true,
+		}).Return(amqp.Queue{Name: queueName}, nil)
```

Comment:
any reason the bools and the amqp.Table arguments aren't higher-scope constants? seems like they're reused across all Push tests

Reply:
I guess there are 2 things here:

1. There is a set of package-scoped constants/variables inside `publisher.go` that are being used. If we reuse them, then we don't really test _an expectation_ of what they should be. I don't think you're commenting on this but it is worth mentioning.

 https://github.com/dominodatalab/forge/blob/27c6d35cbe7948ed3630d80a51f0c03b24ca6a6d/internal/message/amqp/publisher.go#L13-L31

2. To address your question specifically, these tests are _pretty heavy_ in terms of their mental cost for new readers. I think keeping variables localized, albeit duplicative, makes it easier to see how the test works. And, honestly, I'm thought about removing ALL the top-level/shared variables to lower the cost here; I may do that at a later date if I come back to these tests and they scramble my eggs.